File: box-and-box/norm.mjs
// norm.mjs — Deontic Arithmetic, faithful runtime (v0.3)
// The third rung: what OUGHT to be. A deontic status lives in a diamond lattice
//        CONFLICT                 (over-constrained: obligatory ∧ forbidden)
//        /      \
//  OBLIGATORY  FORBIDDEN          (incomparable middles)
//        \      /
//        OPTIONAL                 (permitted, no constraint — the identity)
// accrue = join (commutative, associative, idempotent monoid; identity OPTIONAL,
// absorbing CONFLICT). resolve clears a conflict by priority. detach is factual
// detachment of a conditional norm — and is PARTIAL for contrary-to-duty repairs:
// a repair obligation detaches only AFTER the primary norm is violated. comply is
// the gate; escalate produces the CTD repair. Laws D1–D8 (see test/laws.mjs).

export const STATUS = { OPTIONAL: 'optional', OBLIGATORY: 'obligatory', FORBIDDEN: 'forbidden', CONFLICT: 'conflict' };
const RANK = { optional: 0, obligatory: 1, forbidden: 1, conflict: 2 }; // for monotonicity (middles tie)
export const rank = (s) => RANK[s];
const MOD2STATUS = { obligatory: STATUS.OBLIGATORY, forbidden: STATUS.FORBIDDEN, permitted: STATUS.OPTIONAL };

// join : least upper bound on the diamond lattice
export function join(a, b) {
  if (a === b) return a;
  if (a === STATUS.OPTIONAL) return b;
  if (b === STATUS.OPTIONAL) return a;
  if (a === STATUS.CONFLICT || b === STATUS.CONFLICT) return STATUS.CONFLICT;
  return STATUS.CONFLICT; // {obligatory} ⊔ {forbidden}
}

// a Norm: a conditional rule of one modality, with a priority and an optional CTD repair
export const Norm = (p = {}) => ({
  id: p.id ?? 'norm',
  modality: p.modality ?? 'permitted',     // 'obligatory' | 'forbidden' | 'permitted'
  condition: p.condition ?? (() => true),  // ctx → bool   (factual detachment)
  priority: p.priority ?? 0,               // higher overrides lower in a conflict
  ctd: p.ctd ?? null,                       // Norm — contrary-to-duty repair (in force iff violated)
  target: p.target ?? null
});

const safeCond = (n, ctx) => { try { return !!n.condition(ctx); } catch { return false; } };

// accrue every applicable norm's status into a single verdict (join), tracking contributors
export function adjudicateStatus(ctx, norms) {
  let status = STATUS.OPTIONAL;
  const contributors = [];
  for (const n of norms) {
    if (!safeCond(n, ctx)) continue;
    contributors.push({ id: n.id, modality: n.modality, priority: n.priority });
    status = join(status, MOD2STATUS[n.modality]);
  }
  return { status, contributors };
}

// resolve : clear a CONFLICT by priority (idempotent; identity on a non-conflict verdict)
export function resolve(verdict) {
  if (verdict.status !== STATUS.CONFLICT)
    return { ...verdict, resolved: verdict.status, overridden: [], note: null };
  const ob = verdict.contributors.filter((c) => c.modality === 'obligatory');
  const fb = verdict.contributors.filter((c) => c.modality === 'forbidden');
  const maxOb = Math.max(-Infinity, ...ob.map((c) => c.priority));
  const maxFb = Math.max(-Infinity, ...fb.map((c) => c.priority));
  if (maxOb === maxFb)
    return { ...verdict, resolved: STATUS.CONFLICT, overridden: [], note: 'deadlock: equal priority → escalate' };
  const winnerObligatory = maxOb > maxFb;
  const loser = (winnerObligatory ? fb : ob).map((c) => c.id);
  return {
    ...verdict,
    status: winnerObligatory ? STATUS.OBLIGATORY : STATUS.FORBIDDEN, // makes resolve idempotent
    resolved: winnerObligatory ? STATUS.OBLIGATORY : STATUS.FORBIDDEN,
    overridden: loser,
    note: `${winnerObligatory ? 'obligatory' : 'forbidden'} (p${Math.max(maxOb, maxFb)}) overrides [${loser.join(', ')}]`
  };
}

// detach : factual detachment. A norm is in force when its condition holds. A CTD
// repair detaches ONLY after the primary is violated (partial — like invariant chain).
export function detach(norm, ctx, { violated = false } = {}) {
  return { inForce: safeCond(norm, ctx), repair: (violated && norm.ctd) ? norm.ctd : null };
}

// comply : the gate. Does performing (or omitting) the governed action satisfy its status?
export function comply(status, intend) {
  const violations = [];
  if (status === STATUS.FORBIDDEN && intend) violations.push('performing a forbidden action');
  if (status === STATUS.OBLIGATORY && !intend) violations.push('omitting an obligatory action');
  if (status === STATUS.CONFLICT) violations.push('unresolved normative conflict');
  return { ok: violations.length === 0, violations };
}

// escalate : produce the contrary-to-duty repair obligation now in force
export function escalate(norm, ctx) {
  if (norm && norm.ctd) return { repair: norm.ctd, reason: `CTD: ${norm.id} violated → ${norm.ctd.id} in force` };
  return { repair: Norm({ id: 'escalate-to-human', modality: 'obligatory', priority: Infinity }),
           reason: `${norm ? norm.id : 'obligation'} violated, no CTD → default escalation` };
}

export const statusDigest = (v) => `${v.resolved ?? v.status}${v.overridden && v.overridden.length ? ` (overrode ${v.overridden.join(',')})` : ''}`;

export default { STATUS, rank, join, Norm, adjudicateStatus, resolve, detach, comply, escalate, statusDigest };

File: box-and-box/ports/elixir/lib/box_and_box/norm.ex
defmodule BoxAndBox.Norm do
  @moduledoc """
  Deontic Arithmetic — faithful port of norm.mjs (v0.3).

  Statuses live in a diamond lattice: OPTIONAL ⊑ {OBLIGATORY, FORBIDDEN} ⊑ CONFLICT.
  join is the lub; resolve clears a conflict by priority; detach is factual detachment
  (partial for CTD repairs); comply is the gate; escalate produces the CTD repair.
  Laws D1–D9.
  """

  @optional "optional"
  @obligatory "obligatory"
  @forbidden "forbidden"
  @conflict "conflict"

  def status, do: %{optional: @optional, obligatory: @obligatory, forbidden: @forbidden, conflict: @conflict}

  @rank %{"optional" => 0, "obligatory" => 1, "forbidden" => 1, "conflict" => 2}
  def rank(s), do: Map.get(@rank, s)

  @mod2status %{"obligatory" => @obligatory, "forbidden" => @forbidden, "permitted" => @optional}

  # join : least upper bound on the diamond lattice
  def join(a, b) when a == b, do: a
  def join(@optional, b), do: b
  def join(a, @optional), do: a
  def join(@conflict, _), do: @conflict
  def join(_, @conflict), do: @conflict
  # {obligatory} ⊔ {forbidden}
  def join(_, _), do: @conflict

  # a Norm: conditional rule of one modality, with priority and optional CTD repair
  def norm(p \\ %{}) do
    %{
      id: Map.get(p, :id, "norm"),
      modality: Map.get(p, :modality, "permitted"),
      condition: Map.get(p, :condition, fn _ -> true end),
      priority: Map.get(p, :priority, 0),
      ctd: Map.get(p, :ctd, nil),
      target: Map.get(p, :target, nil)
    }
  end

  defp safe_cond(n, ctx) do
    try do
      !!n.condition.(ctx)
    rescue
      _ -> false
    end
  end

  # accrue every applicable norm's status into a single verdict (join)
  def adjudicate_status(ctx, norms) do
    {status, contributors} =
      Enum.reduce(norms, {@optional, []}, fn n, {st, contribs} ->
        if safe_cond(n, ctx) do
          c = %{id: n.id, modality: n.modality, priority: n.priority}
          {join(st, @mod2status[n.modality]), [c | contribs]}
        else
          {st, contribs}
        end
      end)

    %{status: status, contributors: Enum.reverse(contributors)}
  end

  # resolve : clear a CONFLICT by priority (idempotent; identity on non-conflict)
  def resolve(verdict) do
    if verdict.status != @conflict do
      verdict
      |> Map.put(:resolved, verdict.status)
      |> Map.put(:overridden, [])
      |> Map.put(:note, nil)
    else
      ob = Enum.filter(verdict.contributors, &(&1.modality == "obligatory"))
      fb = Enum.filter(verdict.contributors, &(&1.modality == "forbidden"))
      max_ob = max_priority(ob)
      max_fb = max_priority(fb)

      cond do
        max_ob == max_fb ->
          verdict
          |> Map.put(:resolved, @conflict)
          |> Map.put(:overridden, [])
          |> Map.put(:note, "deadlock: equal priority → escalate")

        true ->
          winner_obligatory = max_ob > max_fb
          loser = (if winner_obligatory, do: fb, else: ob) |> Enum.map(& &1.id)
          win = if winner_obligatory, do: @obligatory, else: @forbidden

          verdict
          |> Map.put(:status, win)
          |> Map.put(:resolved, win)
          |> Map.put(:overridden, loser)
          |> Map.put(
            :note,
            "#{if winner_obligatory, do: "obligatory", else: "forbidden"} (p#{max(max_ob, max_fb)}) overrides [#{Enum.join(loser, ", ")}]"
          )
      end
    end
  end

  # Math.max(-Infinity, ...prios). With no contributors → :neg_infinity sentinel.
  defp max_priority([]), do: :neg_infinity

  defp max_priority(list) do
    Enum.reduce(list, :neg_infinity, fn c, acc -> nmax(acc, c.priority) end)
  end

  defp nmax(:neg_infinity, b), do: b
  defp nmax(a, :neg_infinity), do: a
  defp nmax(a, b), do: max(a, b)

  # detach : factual detachment; CTD repair detaches only after a violation (partial)
  def detach(norm, ctx, opts \\ %{}) do
    violated = Map.get(opts, :violated, false)
    %{in_force: safe_cond(norm, ctx), repair: if(violated and norm.ctd, do: norm.ctd, else: nil)}
  end

  # comply : the gate
  def comply(status, intend) do
    violations =
      []
      |> then(fn v -> if status == @forbidden and intend, do: ["performing a forbidden action" | v], else: v end)
      |> then(fn v -> if status == @obligatory and not intend, do: ["omitting an obligatory action" | v], else: v end)
      |> then(fn v -> if status == @conflict, do: ["unresolved normative conflict" | v], else: v end)

    %{ok: violations == [], violations: Enum.reverse(violations)}
  end

  # escalate : produce the CTD repair obligation now in force
  def escalate(norm, _ctx) do
    if norm && norm.ctd do
      %{repair: norm.ctd, reason: "CTD: #{norm.id} violated → #{norm.ctd.id} in force"}
    else
      %{
        repair: norm(%{id: "escalate-to-human", modality: "obligatory", priority: :infinity}),
        reason: "#{if norm, do: norm.id, else: "obligation"} violated, no CTD → default escalation"
      }
    end
  end
end
